samplers.qmd

```diff
@@ -192,11 +220,111 @@ plot(t(drop(res_hmc$pars)), type = "l", col = "#0000ff33",
 lines(z95[, 1], z95[, 2])
 ```
 
-or viewed over steps:
+Or viewed over steps:
 
 ```{r}
 matplot(t(drop(res_hmc$pars)), lty = 1, type = "l", col = c(2, 4),
         xlab = "Step", ylab = "Value")
 ```
 
-Clearly better!
+Clearly, HMC outperforms the random walk, rapidly exploring the full banana-shaped region with far less random wander.
+
+### Parallel Tempering Sampler
+
+[Parallel tempering](https://en.wikipedia.org/wiki/Parallel_tempering) (also known as replica exchange MCMC) is a technique designed to improve the mixing of Markov chain Monte Carlo methods, particularly in situations where the posterior distribution is multimodal or somehow challenging to explore.
+
+Let's for example define a simple Bayesian model where the likelihood is a mixture model:
+```{r}
+ex_mixture <- function(){
+  fn <- function(x) {
+    log(dnorm(x, mean = 5)/2 + dnorm(x, mean = -5)/2)
+  }
+  monty_model_function(fn,
+                       allow_multiple_parameters = TRUE)
+}
+likelihood <- ex_mixture()
+```
+
+the prior a normal distribution with a wider variance
+
+```{r}
+prior <- monty_dsl({
+  x ~ Normal(0, 10)
+})
+```
+
+And the posterior simply the product of the two (or the sum if working with log-densities)
+```{r}
+posterior <- likelihood + prior
+```
+
+```{r}
+x <- seq(-10, 10, length.out = 1001)
+y <- exp(posterior$density(rbind(x)))
+plot(x, y / sum(y) / diff(x)[[1]], col = "red", type = 'l', ylab = "density")
+```
+
+```{r}
+vcv <- matrix(0.1)
+sampler_rw <- monty_sampler_random_walk(vcv = vcv)
+```
+
+Once the sampler is built, the generic `monty_sample()` function can be used to generate samples:
+
+```{r}
+res_rw <- monty_sample(posterior, sampler_rw, n_steps = 1000)
+```
+
+```{r}
+plot(res_rw$pars[1,,],
+     ylim = c(-8, 8),
+     ylab = "Value",
+     main = "RW samples")
+abline(h = -5, lty = 3, col = "red")
+abline(h = 5, lty = 3, col = "red")
+```
+
+As we can see, the RW sampler does not manage to move outside of the mode where it is initially locked. MCMC theory tells us that it will eventually reach the other side of the distribution, whether with a (low probability) large jump or by accepting several consecutive small disadvantageous steps in the right direction. In practice, it can mean that the second mode might never be explored in the finite amount of time allowed for sampling.
```

Comment:
```suggestion
As we can see, the RW sampler does not manage to move outside of the mode that it starts from. MCMC theory tells us that it will eventually reach the other side of the distribution, whether with a (low probability) large jump or by accepting several consecutive small disadvantageous steps in the right direction. In practice, it can mean that the second mode might never be explored in the finite amount of time allowed for sampling.
```

Reply:
Do we also need to say something about why just running multiple chains from different starting points is unlikely to resolve the issue?